File: ts/src/orderBook.ts
import { AskBook } from "./askBook";
import { BidBook } from "./bidBook";
import { Order } from "./order";

export interface OrderModification {
  price?: number;
  quantity?: number;
}

export class OrderBook {
  private readonly bids = new BidBook();
  private readonly asks = new AskBook();

  addOrder(order: Order): void {
    if (this.findOrder(order.id) !== undefined) {
      throw new Error(`Duplicate order ID: ${order.id}`);
    }

    if (order.side === "buy") {
      this.bids.addOrder(order);
      return;
    }

    this.asks.addOrder(order);
  }

  bestBid(): number | undefined {
    return this.bids.bestPrice();
  }

  bestAsk(): number | undefined {
    return this.asks.bestPrice();
  }

  bestBidOrder(): Order | undefined {
    return this.bids.bestOrder();
  }

  bestAskOrder(): Order | undefined {
    return this.asks.bestOrder();
  }

  dequeueBestBidOrder(): Order | undefined {
    return this.bids.dequeueBestOrder();
  }

  dequeueBestAskOrder(): Order | undefined {
    return this.asks.dequeueBestOrder();
  }

  removeOrder(order: Order): boolean {
    return order.side === "buy"
      ? this.bids.removeOrder(order)
      : this.asks.removeOrder(order);
  }

  reduceOrderQuantity(order: Order, quantity: number): boolean {
    return order.side === "buy"
      ? this.bids.reduceOrderQuantity(order, quantity)
      : this.asks.reduceOrderQuantity(order, quantity);
  }

  findOrder(orderId: string): Order | undefined {
    return this.bids.findOrder(orderId) ?? this.asks.findOrder(orderId);
  }

  cancelOrder(orderId: string): boolean {
    const order = this.findOrder(orderId);
    return order === undefined ? false : this.removeOrder(order);
  }

  modifyOrder(orderId: string, changes: OrderModification): boolean {
    const order = this.findOrder(orderId);
    if (order === undefined) return false;

    const price = changes.price ?? order.price!;
    const quantity = changes.quantity ?? order.quantity;

    if (price <= 0) throw new Error("Modified price must be greater than 0");
    if (quantity <= 0) {
      throw new Error("Modified quantity must be greater than 0");
    }

    const priceChanged = price !== order.price;
    const quantityIncreased = quantity > order.quantity;

    if (!priceChanged && !quantityIncreased) {
      if (quantity < order.quantity) {
        return this.reduceOrderQuantity(order, order.quantity - quantity);
      }
      return true;
    }

    if (!this.removeOrder(order)) {
      throw new Error(`Order ${order.id} could not be removed for modification`);
    }

    this.addOrder({
      ...order,
      price,
      quantity,
      timestamp: Date.now(),
    });
    return true;
  }
}

File: cpp/src/OrderBook.cpp
#include "OrderBook.h"

#include <stdexcept>
#include <utility>

namespace quotient {

Order& OrderBook::add(Order order) {
  validate_order(order);
  if (find(order.id) != nullptr) {
    throw std::invalid_argument("duplicate order id: " + order.id);
  }
  return order.side == Side::Buy ? bids_.add(std::move(order))
                                 : asks_.add(std::move(order));
}

Order* OrderBook::find(const std::string& order_id) {
  if (const Order* order = bids_.find(order_id); order != nullptr) {
    return const_cast<Order*>(order);
  }
  return const_cast<Order*>(asks_.find(order_id));
}

Order* OrderBook::best_bid_order() { return bids_.best_order(); }
Order* OrderBook::best_ask_order() { return asks_.best_order(); }
std::optional<double> OrderBook::best_bid() { return bids_.best_price(); }
std::optional<double> OrderBook::best_ask() { return asks_.best_price(); }

bool OrderBook::remove(const Order& order) {
  return order.side == Side::Buy ? bids_.remove(order) : asks_.remove(order);
}

bool OrderBook::reduce(const Order& order, std::uint64_t quantity) {
  return order.side == Side::Buy ? bids_.reduce(order, quantity)
                                 : asks_.reduce(order, quantity);
}

bool OrderBook::cancel(const std::string& order_id) {
  const Order* order = find(order_id);
  if (order == nullptr) return false;
  const Order copy = *order;
  return remove(copy);
}

bool OrderBook::modify(const std::string& order_id,
                       const OrderModification& changes) {
  Order* order = find(order_id);
  if (order == nullptr) return false;
  const double price = changes.price.value_or(*order->price);
  const std::uint64_t quantity = changes.quantity.value_or(order->quantity);
  if (price <= 0 || quantity == 0) {
    throw std::invalid_argument("modified price and quantity must be positive");
  }

  const bool price_changed = price != *order->price;
  const bool quantity_increased = quantity > order->quantity;
  if (!price_changed && !quantity_increased) {
    if (quantity < order->quantity) return reduce(*order, order->quantity - quantity);
    return true;
  }

  Order replacement = *order;
  if (!remove(replacement)) throw std::runtime_error("failed to remove order");
  replacement.price = price;
  replacement.quantity = quantity;
  replacement.timestamp = now_millis();
  add(std::move(replacement));
  return true;
}

BookDepth OrderBook::depth() const { return {bids_.depth(), asks_.depth()}; }

}  // namespace quotient
